src/render/template.rs

```diff
@@ -280,6 +282,132 @@ impl HelperDef for VersionCheckHelper {
+            "%" => Some(a % b),
+            _ => None,
+        }
+    }
+
+    fn new() -> Box<MathHelper> {
+        Box::new(MathHelper {})
+    }
+}
+
+/**
+Simple math helper, which can do the usual arithmetic operations on integers and floats. Tries to handle most edge-cases and provide useful error message to the user.
+
+Usage: `{{ math 5 "+" 3 }}`, `{{ math 23.8 / -1.5}}`, `{{ math "3" "*" 8.5 }}`
+
+Supported operations:
+   - \+ addition
+   - \- subtraction
+   - \* multiplication
+   - / **decimal** division (integers are converted to floats)
+   - // integer division (both numbers must be integers)
+   - % modulo
+   - & bitwise and (integers only)
+   - | bitwise or (integers only)
+   - ^ bitwise xor (integers only)
+   - << bitwise shift left (integers only)
+   - \>> bitwise shift right (integers only)
+ */
+impl HelperDef for MathHelper {
+    fn call_inner<'reg: 'rc, 'rc>(
+        &self,
+        h: &hb::Helper<'reg, 'rc>,
+        _: &'reg Handlebars<'reg>,
+        _: &'rc hb::Context,
+        _: &mut hb::RenderContext<'reg, 'rc>,
+    ) -> Result<hb::ScopedJson<'reg, 'rc>, RenderError> {
+        let wrong_param_count = format!("math: Found {} parameters, but math helper requires 3 parameters: number, operator as a string, number. Example: {}.", h.params().len(), "{{ math 1 \"+\" 2.5 }}");
+
+        let a = h.param(0).ok_or(RenderError::new(&wrong_param_count))?;
+        let operation = h.param(1).ok_or(RenderError::new(&wrong_param_count))?;
+        let b = h.param(2).ok_or(RenderError::new(&wrong_param_count))?;
+        let operation = operation.value().as_str().ok_or(RenderError::new(
+            "math: Second argument must be a string. Example: {{ math 1 \"+\" 2 }}.",
+        ))?;
+
+        let aint = a
+            .value()
+            .as_i64()
+            .or(a.value().as_str().and_then(|s| i64::from_str(s).ok()));
+        let afloat = a
+            .value()
+            .as_f64()
+            .or(a.value().as_str().and_then(|s| f64::from_str(s).ok()));
+        let bint = b
+            .value()
+            .as_i64()
+            .or(b.value().as_str().and_then(|s| i64::from_str(s).ok()));
+        let bfloat = b
+            .value()
+            .as_f64()
+            .or(b.value().as_str().and_then(|s| f64::from_str(s).ok()));
+
+        // try integer arithmetics
+        if let (Some(aint), Some(bint)) = (aint, bint) {
+            if operation != "/" {
+                // normal division is done using floats to make it simpler for inexperienced users. For integer division, use //.
+                return if let Some(r) = MathHelper::hb_math_int(aint, operation, bint) {
+                    Ok(hb::ScopedJson::Derived(JsonValue::Number(Number::from(r))))
+                } else {
+                    Err(RenderError::new(format!("math: Operation \"{}\" is not possible with integers. Available operations on integers: +, -, *, /, //, %, &, |, ^, <<, >>", operation)))
```

Comment:
That would move the error creation into `hb_math_int()`, but currently all the errors (and their descriptions) are created in the same function making it easier to keep track of them. Also `call_inner()` is the function taking care of all the possible results, so it makes sense that the final form of the error is created in `call_inner()` function and not hidden somewhere else.

Reply:
If a code is extracted into a helper function it should be extracted with error handling as well. It's not going to be very far / hidden, it's the same type. Since in Rust error passing is always explicit, it should be clear where the error is coming from.
Pls also use `Self::` when calling function from inside the type, ie. `Self::hb_math_int(...)`.

Reply:
Ok, done.